nifi-nar-bundles/nifi-iceberg-bundle/nifi-iceberg-processors/src/main/java/org/apache/nifi/processors/iceberg/converter/GenericDataConverters.java

```diff
@@ -48,10 +51,28 @@
 public class GenericDataConverters {
 
     static class SameTypeConverter extends DataConverter<Object, Object> {
+        final Type.PrimitiveType targetType;
+        final DataType sourceType;
+
+        public SameTypeConverter(final Type.PrimitiveType type, final DataType dataType) {
+            targetType = type;
+            sourceType = dataType;
+        }
 
         @Override
         public Object convert(Object data) {
-            return data;
+            switch (targetType.typeId()) {
+                case BOOLEAN: return DataTypeUtils.convertType(data, RecordFieldType.BOOLEAN.getDataType(), null);
+                case INTEGER: return DataTypeUtils.convertType(data, RecordFieldType.INT.getDataType(), null);
+                case LONG: return DataTypeUtils.convertType(data, RecordFieldType.LONG.getDataType(), null);
+                case FLOAT: return DataTypeUtils.convertType(data, RecordFieldType.FLOAT.getDataType(), null);
+                case DOUBLE: return DataTypeUtils.convertType(data, RecordFieldType.DOUBLE.getDataType(), null);
+                // DATE objects are already LocalDates, so return it untouched
+                case DATE: return data;
+                case STRING:
+                default:
+                    return DataTypeUtils.convertRecordFieldtoObject(data, RecordFieldType.STRING.getDataType());
```

Comment:
The UUID data type is also using the `SameTypeConverter` and since it is not handled in the switch case it is going to be converted to String instead of UUID.